web_src/js/features/repo-diff.ts

```diff
@@ -219,30 +222,51 @@ function initRepoDiffShowMore() {
 async function onLocationHashChange() {
   // try to scroll to the target element by the current hash
   const currentHash = window.location.hash;
-  if (!currentHash.startsWith('#diff-') && !currentHash.startsWith('#issuecomment-')) return;
+  const issueCommentPrefix = '#issuecomment-';
+  const isDiffHash = currentHash.startsWith('#diff-');
+  const isIssueCommentHash = currentHash.startsWith(issueCommentPrefix);
+  if (!isDiffHash && !isIssueCommentHash) return;
 
   // avoid reentrance when we are changing the hash to scroll and trigger ":target" selection
-  const attrAutoScrollRunning = 'data-auto-scroll-running';
-  if (document.body.hasAttribute(attrAutoScrollRunning)) return;
+  if (document.body.hasAttribute(diffAutoScrollAttr)) return;
 
-  const targetElementId = currentHash.substring(1);
-  while (currentHash === window.location.hash) {
-    // use getElementById to avoid querySelector throws an error when the hash is invalid
-    // eslint-disable-next-line unicorn/prefer-query-selector
-    const targetElement = document.getElementById(targetElementId);
-    if (targetElement) {
-      // need to change hash to re-trigger ":target" CSS selector, let's manually scroll to it
-      targetElement.scrollIntoView();
-      document.body.setAttribute(attrAutoScrollRunning, 'true');
-      window.location.hash = '';
-      window.location.hash = currentHash;
-      setTimeout(() => document.body.removeAttribute(attrAutoScrollRunning), 0);
+  const hashValue = currentHash.substring(1);
+  let targetElementId = hashValue;
+
+  if (isDiffHash) {
+    const success = await highlightDiffSelectionFromHash();
+    if (success) {
+      // Successfully highlighted and scrolled, we're done
       return;
     }
+    const range = parseDiffHashRange(hashValue);
+    if (range) {
+      targetElementId = `${range.fragment}${range.startSide}${range.startLine}`;
+    }
+  }
+
+  while (currentHash === window.location.hash) {
+    if (isDiffHash) {
+      // eslint-disable-next-line unicorn/prefer-query-selector
+      const targetElement = document.getElementById(targetElementId);
+      if (targetElement) {
+        // Try again to highlight and scroll now that the element is loaded
+        const success = await highlightDiffSelectionFromHash();
+        if (success) return;
```

Comment:
Is this loop still right?

If the selected line are in the expanded area (blob excerpt), now you just keep "loading more files", but the selected range can never be really loaded.